sklearn/metrics/pairwise.py

```diff
@@ -1572,13 +1577,28 @@ def _parallel_pairwise(X, Y, func, n_jobs, **kwds):
     if effective_n_jobs(n_jobs) == 1:
         return func(X, Y, **kwds)
 
-    # enforce a threading backend to prevent data communication overhead
-    fd = delayed(_dist_wrapper)
-    ret = np.empty((X.shape[0], Y.shape[0]), dtype=dtype, order="F")
-    Parallel(backend="threading", n_jobs=n_jobs)(
-        fd(func, ret, s, X, Y[s], **kwds)
-        for s in gen_even_slices(_num_samples(Y), effective_n_jobs(n_jobs))
-    )
+    # preferring a threading backend to prevent data communication overhead
+    active_backend, _ = get_active_backend(prefer="threads")
+    uses_threads = getattr(active_backend, 'uses_threads', False)
+
+    if uses_threads:
+        # adopting a faster write-in-place approach
+        fd = delayed(_dist_wrapper_mlthrd)
+        ret = np.empty((X.shape[0], Y.shape[0]), dtype=dtype, order="F")
+        Parallel(backend="threading", n_jobs=n_jobs)(
+            fd(func, ret, s, X, Y[s], **kwds)
+            for s in gen_even_slices(_num_samples(Y), effective_n_jobs(n_jobs))
+        )
+    else:
+        # each process returns its chunk of data
+        fd = delayed(_dist_wrapper_mltprc)
+        ret = np.empty((X.shape[0], Y.shape[0]), dtype=dtype, order="F")
+        distchunks = Parallel(n_jobs=n_jobs)(
+            fd(func, s, X, Y[s], **kwds) for s in gen_even_slices(_num_samples(Y), effective_n_jobs(n_jobs))
+        )
+        for slice_, chunk in zip(gen_even_slices(_num_samples(Y), effective_n_jobs(n_jobs)), distchunks):
+            ret[:, slice_] = chunk
```

Comment:
I suggest:

```python
    ret = np.empty((X.shape[0], Y.shape[0]), dtype=dtype, order="F")
    if uses_threads:
        # adopting a faster write-in-place approach
        fd = delayed(_dist_wrapper)
        Parallel(backend="threading", n_jobs=n_jobs)(
            fd(func, ret, s, X, Y[s], **kwds)
            for s in gen_even_slices(_num_samples(Y), effective_n_jobs(n_jobs))
        )
    else:
        # each process returns its chunk of data
        fd = delayed(func)
        even_slices = list(gen_even_slices(_num_samples(Y), effective_n_jobs(n_jobs)))
        distchunks = Parallel(n_jobs=n_jobs)(fd(X, Y[s], **kwds) for s in even_slices)
        for slice_, chunk in zip(even_slices, distchunks):
            ret[:, slice_] = chunk
```

Details:

1. `ret` can be shared with both branches
2. For multi-processing `fd` can be set to `delayed(func)` so we do not need another callable for multiprocessing
3. Since we are using the slices twice, we can hold on it to and use it twice.